homeassistant/components/mqtt/discovery.py

```diff
@@ -187,27 +185,31 @@ async def async_discovery_message_received(msg):
 
         await async_process_discovery_payload(component, discovery_id, payload)
 
-    async def async_process_discovery_payload(component, discovery_id, payload):
+    async def async_process_discovery_payload(
+        component: str, discovery_id: str, payload: ConfigType
+    ) -> None:
         """Process the payload of a new discovery."""
 
         _LOGGER.debug("Process discovery payload %s", payload)
         discovery_hash = (component, discovery_id)
-        if discovery_hash in hass.data[ALREADY_DISCOVERED] or payload:
+        if discovery_hash in mqtt_data.discovery_already_discovered or payload:
 
-            async def discovery_done(_):
-                pending = hass.data[PENDING_DISCOVERED][discovery_hash]["pending"]
+            async def discovery_done(_) -> None:
+                pending = mqtt_data.discovery_pending_discovered[discovery_hash][
+                    "pending"
+                ]
                 _LOGGER.debug("Pending discovery for %s: %s", discovery_hash, pending)
                 if not pending:
-                    hass.data[PENDING_DISCOVERED][discovery_hash]["unsub"]()
-                    hass.data[PENDING_DISCOVERED].pop(discovery_hash)
+                    mqtt_data.discovery_pending_discovered[discovery_hash]["unsub"]()
+                    mqtt_data.discovery_pending_discovered.pop(discovery_hash)
```

Comment:
Code style here: would it make sense to first `pop`, and then run `unsub` on the result?